// @thirdweb-dev/sdk/dist/thirdweb-dev-sdk.cjs.dev.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var dist_thirdwebDevSdk = require('./index-a3d6bc85.cjs.dev.js');
require('ethers');
require('zod');
require('@thirdweb-dev/storage');
require('uuid');
require('eventemitter3');
require('@thirdweb-dev/contracts-js/dist/abis/IContractMetadata.json');
require('@thirdweb-dev/contracts-js/dist/abis/IPermissions.json');
require('@thirdweb-dev/contracts-js/dist/abis/IPlatformFee.json');
require('@thirdweb-dev/contracts-js/dist/abis/IPrimarySale.json');
require('@thirdweb-dev/contracts-js/dist/abis/IRoyalty.json');
require('@thirdweb-dev/contracts-js/dist/abis/IBurnableERC20.json');
require('@thirdweb-dev/contracts-js/dist/abis/IDropSinglePhase.json');
require('@thirdweb-dev/contracts-js/dist/abis/IERC20.json');
require('@thirdweb-dev/contracts-js/dist/abis/IMintableERC20.json');
require('@thirdweb-dev/contracts-js/dist/abis/IMulticall.json');
require('@thirdweb-dev/contracts-js/dist/abis/ISignatureMintERC20.json');
require('@thirdweb-dev/contracts-js/dist/abis/IBurnableERC721.json');
require('@thirdweb-dev/contracts-js/dist/abis/IClaimableERC721.json');
require('@thirdweb-dev/contracts-js/dist/abis/IDelayedReveal.json');
require('@thirdweb-dev/contracts-js/dist/abis/IERC721.json');
require('@thirdweb-dev/contracts-js/dist/abis/IERC721Enumerable.json');
require('@thirdweb-dev/contracts-js/dist/abis/IERC721Supply.json');
require('@thirdweb-dev/contracts-js/dist/abis/ILazyMint.json');
require('@thirdweb-dev/contracts-js/dist/abis/IMintableERC721.json');
require('@thirdweb-dev/contracts-js/dist/abis/SignatureMintERC721.json');
require('@thirdweb-dev/contracts-js/dist/abis/DelayedReveal.json');
require('@thirdweb-dev/contracts-js/dist/abis/DropSinglePhase1155.json');
require('@thirdweb-dev/contracts-js/dist/abis/IBurnableERC1155.json');
require('@thirdweb-dev/contracts-js/dist/abis/IClaimableERC1155.json');
require('@thirdweb-dev/contracts-js/dist/abis/IERC1155.json');
require('@thirdweb-dev/contracts-js/dist/abis/IERC1155Enumerable.json');
require('@thirdweb-dev/contracts-js/dist/abis/IMintableERC1155.json');
require('@thirdweb-dev/contracts-js/dist/abis/ISignatureMintERC1155.json');
require('tiny-invariant');
require('@thirdweb-dev/contracts-js/dist/abis/IERC165.json');
require('@thirdweb-dev/contracts-js/dist/abis/IERC721Metadata.json');
require('@thirdweb-dev/contracts-js/dist/abis/IERC1155Metadata.json');
require('@thirdweb-dev/contracts-js/dist/abis/IDelayedRevealDeprecated.json');
require('cross-fetch');
require('@thirdweb-dev/contracts-js/dist/abis/Forwarder.json');
require('@thirdweb-dev/contracts-js/dist/abis/IERC20Metadata.json');
require('fast-deep-equal');
require('@thirdweb-dev/contracts-js/dist/abis/ContractPublisher.json');
require('@thirdweb-dev/contracts-js/dist/abis/IThirdwebContract.json');
require('@thirdweb-dev/contracts-js/dist/abis/TWFactory.json');
require('@thirdweb-dev/contracts-js/dist/abis/TWRegistry.json');



exports.ALL_ROLES = dist_thirdwebDevSdk.ALL_ROLES;
exports.AbiObjectSchema = dist_thirdwebDevSdk.AbiObjectSchema;
exports.AbiSchema = dist_thirdwebDevSdk.AbiSchema;
exports.AbiTypeSchema = dist_thirdwebDevSdk.AbiTypeSchema;
exports.AdminRoleMissingError = dist_thirdwebDevSdk.AdminRoleMissingError;
exports.AssetNotFoundError = dist_thirdwebDevSdk.AssetNotFoundError;
exports.AuctionAlreadyStartedError = dist_thirdwebDevSdk.AuctionAlreadyStartedError;
exports.AuctionHasNotEndedError = dist_thirdwebDevSdk.AuctionHasNotEndedError;
exports.AuthenticationOptionsSchema = dist_thirdwebDevSdk.AuthenticationOptionsSchema;
exports.AuthenticationPayloadDataSchema = dist_thirdwebDevSdk.AuthenticationPayloadDataSchema;
exports.AuthenticationPayloadSchema = dist_thirdwebDevSdk.AuthenticationPayloadSchema;
exports.BYOCContractMetadataSchema = dist_thirdwebDevSdk.BYOCContractMetadataSchema;
exports.BaseSignaturePayloadInput = dist_thirdwebDevSdk.BaseSignaturePayloadInput;
exports.CONTRACTS_MAP = dist_thirdwebDevSdk.CONTRACTS_MAP;
exports.CONTRACT_ADDRESSES = dist_thirdwebDevSdk.CONTRACT_ADDRESSES;
Object.defineProperty(exports, 'ChainId', {
	enumerable: true,
	get: function () { return dist_thirdwebDevSdk.ChainId; }
});
exports.ChainIdToAddressSchema = dist_thirdwebDevSdk.ChainIdToAddressSchema;
exports.ClaimConditionInputArray = dist_thirdwebDevSdk.ClaimConditionInputArray;
exports.ClaimConditionInputSchema = dist_thirdwebDevSdk.ClaimConditionInputSchema;
exports.ClaimConditionOutputSchema = dist_thirdwebDevSdk.ClaimConditionOutputSchema;
Object.defineProperty(exports, 'ClaimEligibility', {
	enumerable: true,
	get: function () { return dist_thirdwebDevSdk.ClaimEligibility; }
});
exports.CommonContractOutputSchema = dist_thirdwebDevSdk.CommonContractOutputSchema;
exports.CommonContractSchema = dist_thirdwebDevSdk.CommonContractSchema;
exports.CommonPlatformFeeSchema = dist_thirdwebDevSdk.CommonPlatformFeeSchema;
exports.CommonPrimarySaleSchema = dist_thirdwebDevSdk.CommonPrimarySaleSchema;
exports.CommonRoyaltySchema = dist_thirdwebDevSdk.CommonRoyaltySchema;
exports.CommonSymbolSchema = dist_thirdwebDevSdk.CommonSymbolSchema;
exports.CommonTrustedForwarderSchema = dist_thirdwebDevSdk.CommonTrustedForwarderSchema;
exports.CompilerMetadataFetchedSchema = dist_thirdwebDevSdk.CompilerMetadataFetchedSchema;
exports.ContractDeployer = dist_thirdwebDevSdk.ContractDeployer;
exports.ContractEncoder = dist_thirdwebDevSdk.ContractEncoder;
exports.ContractEvents = dist_thirdwebDevSdk.ContractEvents;
exports.ContractInfoSchema = dist_thirdwebDevSdk.ContractInfoSchema;
exports.ContractInterceptor = dist_thirdwebDevSdk.ContractInterceptor;
exports.ContractMetadata = dist_thirdwebDevSdk.ContractMetadata;
exports.ContractPlatformFee = dist_thirdwebDevSdk.ContractPlatformFee;
exports.ContractPrimarySale = dist_thirdwebDevSdk.ContractPrimarySale;
exports.ContractPublishedMetadata = dist_thirdwebDevSdk.ContractPublishedMetadata;
exports.ContractRoles = dist_thirdwebDevSdk.ContractRoles;
exports.ContractRoyalty = dist_thirdwebDevSdk.ContractRoyalty;
exports.CurrencySchema = dist_thirdwebDevSdk.CurrencySchema;
exports.CurrencyValueSchema = dist_thirdwebDevSdk.CurrencyValueSchema;
exports.CustomContractDeploy = dist_thirdwebDevSdk.CustomContractDeploy;
exports.CustomContractInput = dist_thirdwebDevSdk.CustomContractInput;
exports.CustomContractOutput = dist_thirdwebDevSdk.CustomContractOutput;
exports.CustomContractSchema = dist_thirdwebDevSdk.CustomContractSchema;
exports.DEFAULT_IPFS_GATEWAY = dist_thirdwebDevSdk.DEFAULT_IPFS_GATEWAY;
exports.DEFAULT_QUERY_ALL_COUNT = dist_thirdwebDevSdk.DEFAULT_QUERY_ALL_COUNT;
exports.DelayedReveal = dist_thirdwebDevSdk.DelayedReveal;
exports.DropClaimConditions = dist_thirdwebDevSdk.DropClaimConditions;
exports.DropErc1155ClaimConditions = dist_thirdwebDevSdk.DropErc1155ClaimConditions;
exports.DropErc1155History = dist_thirdwebDevSdk.DropErc1155History;
exports.DuplicateFileNameError = dist_thirdwebDevSdk.DuplicateFileNameError;
exports.DuplicateLeafsError = dist_thirdwebDevSdk.DuplicateLeafsError;
exports.Edition = dist_thirdwebDevSdk.Edition;
exports.EditionDrop = dist_thirdwebDevSdk.EditionDrop;
exports.EditionMetadataInputOrUriSchema = dist_thirdwebDevSdk.EditionMetadataInputOrUriSchema;
exports.EditionMetadataInputSchema = dist_thirdwebDevSdk.EditionMetadataInputSchema;
exports.EditionMetadataOutputSchema = dist_thirdwebDevSdk.EditionMetadataOutputSchema;
exports.EditionMetadataWithOwnerOutputSchema = dist_thirdwebDevSdk.EditionMetadataWithOwnerOutputSchema;
exports.Erc1155 = dist_thirdwebDevSdk.Erc1155;
exports.Erc1155BatchMintable = dist_thirdwebDevSdk.Erc1155BatchMintable;
exports.Erc1155Burnable = dist_thirdwebDevSdk.Erc1155Burnable;
exports.Erc1155Enumerable = dist_thirdwebDevSdk.Erc1155Enumerable;
exports.Erc1155LazyMintable = dist_thirdwebDevSdk.Erc1155LazyMintable;
exports.Erc1155Mintable = dist_thirdwebDevSdk.Erc1155Mintable;
exports.Erc1155SignatureMintable = dist_thirdwebDevSdk.Erc1155SignatureMintable;
exports.Erc20 = dist_thirdwebDevSdk.Erc20;
exports.Erc20BatchMintable = dist_thirdwebDevSdk.Erc20BatchMintable;
exports.Erc20Burnable = dist_thirdwebDevSdk.Erc20Burnable;
exports.Erc20Mintable = dist_thirdwebDevSdk.Erc20Mintable;
exports.Erc20SignatureMintable = dist_thirdwebDevSdk.Erc20SignatureMintable;
exports.Erc721 = dist_thirdwebDevSdk.Erc721;
exports.Erc721BatchMintable = dist_thirdwebDevSdk.Erc721BatchMintable;
exports.Erc721Burnable = dist_thirdwebDevSdk.Erc721Burnable;
exports.Erc721ClaimableWithConditions = dist_thirdwebDevSdk.Erc721ClaimableWithConditions;
exports.Erc721Enumerable = dist_thirdwebDevSdk.Erc721Enumerable;
exports.Erc721LazyMintable = dist_thirdwebDevSdk.Erc721LazyMintable;
exports.Erc721Mintable = dist_thirdwebDevSdk.Erc721Mintable;
exports.Erc721Supply = dist_thirdwebDevSdk.Erc721Supply;
exports.Erc721WithQuantitySignatureMintable = dist_thirdwebDevSdk.Erc721WithQuantitySignatureMintable;
Object.defineProperty(exports, 'EventType', {
	enumerable: true,
	get: function () { return dist_thirdwebDevSdk.EventType; }
});
exports.ExtensionNotImplementedError = dist_thirdwebDevSdk.ExtensionNotImplementedError;
exports.ExtraPublishMetadataSchemaInput = dist_thirdwebDevSdk.ExtraPublishMetadataSchemaInput;
exports.ExtraPublishMetadataSchemaOutput = dist_thirdwebDevSdk.ExtraPublishMetadataSchemaOutput;
exports.FactoryDeploymentSchema = dist_thirdwebDevSdk.FactoryDeploymentSchema;
exports.FetchError = dist_thirdwebDevSdk.FetchError;
exports.FileNameMissingError = dist_thirdwebDevSdk.FileNameMissingError;
exports.FullPublishMetadataSchemaInput = dist_thirdwebDevSdk.FullPublishMetadataSchemaInput;
exports.FullPublishMetadataSchemaOutput = dist_thirdwebDevSdk.FullPublishMetadataSchemaOutput;
exports.FunctionDeprecatedError = dist_thirdwebDevSdk.FunctionDeprecatedError;
exports.GasCostEstimator = dist_thirdwebDevSdk.GasCostEstimator;
exports.InterfaceId_IERC1155 = dist_thirdwebDevSdk.InterfaceId_IERC1155;
exports.InterfaceId_IERC721 = dist_thirdwebDevSdk.InterfaceId_IERC721;
exports.InvalidAddressError = dist_thirdwebDevSdk.InvalidAddressError;
exports.ListingNotFoundError = dist_thirdwebDevSdk.ListingNotFoundError;
Object.defineProperty(exports, 'ListingType', {
	enumerable: true,
	get: function () { return dist_thirdwebDevSdk.ListingType; }
});
exports.LoginOptionsSchema = dist_thirdwebDevSdk.LoginOptionsSchema;
exports.LoginPayloadDataSchema = dist_thirdwebDevSdk.LoginPayloadDataSchema;
exports.LoginPayloadSchema = dist_thirdwebDevSdk.LoginPayloadSchema;
exports.Marketplace = dist_thirdwebDevSdk.Marketplace;
exports.MarketplaceAuction = dist_thirdwebDevSdk.MarketplaceAuction;
exports.MarketplaceDirect = dist_thirdwebDevSdk.MarketplaceDirect;
exports.MerkleSchema = dist_thirdwebDevSdk.MerkleSchema;
exports.MintRequest1155 = dist_thirdwebDevSdk.MintRequest1155;
exports.MintRequest20 = dist_thirdwebDevSdk.MintRequest20;
exports.MintRequest721 = dist_thirdwebDevSdk.MintRequest721;
exports.MintRequest721withQuantity = dist_thirdwebDevSdk.MintRequest721withQuantity;
exports.MissingOwnerRoleError = dist_thirdwebDevSdk.MissingOwnerRoleError;
exports.MissingRoleError = dist_thirdwebDevSdk.MissingRoleError;
exports.Multiwrap = dist_thirdwebDevSdk.Multiwrap;
exports.NATIVE_TOKENS = dist_thirdwebDevSdk.NATIVE_TOKENS;
exports.NATIVE_TOKEN_ADDRESS = dist_thirdwebDevSdk.NATIVE_TOKEN_ADDRESS;
exports.NFTCollection = dist_thirdwebDevSdk.NFTCollection;
exports.NFTDrop = dist_thirdwebDevSdk.NFTDrop;
exports.NotEnoughTokensError = dist_thirdwebDevSdk.NotEnoughTokensError;
exports.NotFoundError = dist_thirdwebDevSdk.NotFoundError;
exports.OZ_DEFENDER_FORWARDER_ADDRESS = dist_thirdwebDevSdk.OZ_DEFENDER_FORWARDER_ADDRESS;
exports.OptionalPropertiesInput = dist_thirdwebDevSdk.OptionalPropertiesInput;
exports.PINATA_IPFS_URL = dist_thirdwebDevSdk.PINATA_IPFS_URL;
exports.PREBUILT_CONTRACTS_MAP = dist_thirdwebDevSdk.PREBUILT_CONTRACTS_MAP;
exports.PUBLIC_GATEWAYS = dist_thirdwebDevSdk.PUBLIC_GATEWAYS;
exports.Pack = dist_thirdwebDevSdk.Pack;
exports.PartialClaimConditionInputSchema = dist_thirdwebDevSdk.PartialClaimConditionInputSchema;
exports.PreDeployMetadata = dist_thirdwebDevSdk.PreDeployMetadata;
exports.PreDeployMetadataFetchedSchema = dist_thirdwebDevSdk.PreDeployMetadataFetchedSchema;
exports.ProfileSchemaInput = dist_thirdwebDevSdk.ProfileSchemaInput;
exports.ProfileSchemaOutput = dist_thirdwebDevSdk.ProfileSchemaOutput;
Object.defineProperty(exports, 'ProposalState', {
	enumerable: true,
	get: function () { return dist_thirdwebDevSdk.ProposalState; }
});
exports.PublishedContractSchema = dist_thirdwebDevSdk.PublishedContractSchema;
exports.QuantityAboveLimitError = dist_thirdwebDevSdk.QuantityAboveLimitError;
exports.QuantitySchema = dist_thirdwebDevSdk.QuantitySchema;
exports.RestrictedTransferError = dist_thirdwebDevSdk.RestrictedTransferError;
exports.SUPPORTED_CHAIN_IDS = dist_thirdwebDevSdk.SUPPORTED_CHAIN_IDS;
exports.Signature1155PayloadInput = dist_thirdwebDevSdk.Signature1155PayloadInput;
exports.Signature1155PayloadInputWithTokenId = dist_thirdwebDevSdk.Signature1155PayloadInputWithTokenId;
exports.Signature1155PayloadOutput = dist_thirdwebDevSdk.Signature1155PayloadOutput;
exports.Signature20PayloadInput = dist_thirdwebDevSdk.Signature20PayloadInput;
exports.Signature20PayloadOutput = dist_thirdwebDevSdk.Signature20PayloadOutput;
exports.Signature721PayloadInput = dist_thirdwebDevSdk.Signature721PayloadInput;
exports.Signature721PayloadOutput = dist_thirdwebDevSdk.Signature721PayloadOutput;
exports.Signature721WithQuantityInput = dist_thirdwebDevSdk.Signature721WithQuantityInput;
exports.Signature721WithQuantityOutput = dist_thirdwebDevSdk.Signature721WithQuantityOutput;
exports.SignatureDrop = dist_thirdwebDevSdk.SignatureDrop;
exports.SnapshotAddressInput = dist_thirdwebDevSdk.SnapshotAddressInput;
exports.SnapshotInfoSchema = dist_thirdwebDevSdk.SnapshotInfoSchema;
exports.SnapshotInputSchema = dist_thirdwebDevSdk.SnapshotInputSchema;
exports.SnapshotSchema = dist_thirdwebDevSdk.SnapshotSchema;
exports.Split = dist_thirdwebDevSdk.Split;
exports.StandardErc1155 = dist_thirdwebDevSdk.StandardErc1155;
exports.StandardErc20 = dist_thirdwebDevSdk.StandardErc20;
exports.StandardErc721 = dist_thirdwebDevSdk.StandardErc721;
exports.TW_IPFS_SERVER_URL = dist_thirdwebDevSdk.TW_IPFS_SERVER_URL;
exports.ThirdwebSDK = dist_thirdwebDevSdk.ThirdwebSDK;
exports.Token = dist_thirdwebDevSdk.Token;
exports.TokenDrop = dist_thirdwebDevSdk.TokenDrop;
exports.TokenERC20History = dist_thirdwebDevSdk.TokenERC20History;
exports.TokenMintInputSchema = dist_thirdwebDevSdk.TokenMintInputSchema;
exports.TransactionError = dist_thirdwebDevSdk.TransactionError;
exports.UploadError = dist_thirdwebDevSdk.UploadError;
exports.UserWallet = dist_thirdwebDevSdk.UserWallet;
exports.VerifyOptionsSchema = dist_thirdwebDevSdk.VerifyOptionsSchema;
exports.Vote = dist_thirdwebDevSdk.Vote;
Object.defineProperty(exports, 'VoteType', {
	enumerable: true,
	get: function () { return dist_thirdwebDevSdk.VoteType; }
});
exports.WalletAuthenticator = dist_thirdwebDevSdk.WalletAuthenticator;
exports.WrongListingTypeError = dist_thirdwebDevSdk.WrongListingTypeError;
exports.assertEnabled = dist_thirdwebDevSdk.assertEnabled;
exports.convertToTWError = dist_thirdwebDevSdk.convertToTWError;
exports.createSnapshot = dist_thirdwebDevSdk.createSnapshot;
exports.detectContractFeature = dist_thirdwebDevSdk.detectContractFeature;
exports.detectFeatures = dist_thirdwebDevSdk.detectFeatures;
exports.extractConstructorParams = dist_thirdwebDevSdk.extractConstructorParams;
exports.extractConstructorParamsFromAbi = dist_thirdwebDevSdk.extractConstructorParamsFromAbi;
exports.extractEventsFromAbi = dist_thirdwebDevSdk.extractEventsFromAbi;
exports.extractFunctionParamsFromAbi = dist_thirdwebDevSdk.extractFunctionParamsFromAbi;
exports.extractFunctions = dist_thirdwebDevSdk.extractFunctions;
exports.extractFunctionsFromAbi = dist_thirdwebDevSdk.extractFunctionsFromAbi;
exports.fetchContractMetadata = dist_thirdwebDevSdk.fetchContractMetadata;
exports.fetchContractMetadataFromAddress = dist_thirdwebDevSdk.fetchContractMetadataFromAddress;
exports.fetchExtendedReleaseMetadata = dist_thirdwebDevSdk.fetchExtendedReleaseMetadata;
exports.fetchPreDeployMetadata = dist_thirdwebDevSdk.fetchPreDeployMetadata;
exports.fetchRawPredeployMetadata = dist_thirdwebDevSdk.fetchRawPredeployMetadata;
exports.fetchSourceFilesFromMetadata = dist_thirdwebDevSdk.fetchSourceFilesFromMetadata;
exports.getContractAddressByChainId = dist_thirdwebDevSdk.getContractAddressByChainId;
exports.getContractPublisherAddress = dist_thirdwebDevSdk.getContractPublisherAddress;
exports.getContractTypeForRemoteName = dist_thirdwebDevSdk.getContractTypeForRemoteName;
exports.getNativeTokenByChainId = dist_thirdwebDevSdk.getNativeTokenByChainId;
exports.getProviderForNetwork = dist_thirdwebDevSdk.getProviderForNetwork;
exports.getReadOnlyProvider = dist_thirdwebDevSdk.getReadOnlyProvider;
exports.getRoleHash = dist_thirdwebDevSdk.getRoleHash;
exports.hasFunction = dist_thirdwebDevSdk.hasFunction;
exports.hashLeafNode = dist_thirdwebDevSdk.hashLeafNode;
exports.includesErrorMessage = dist_thirdwebDevSdk.includesErrorMessage;
exports.isDowngradeVersion = dist_thirdwebDevSdk.isDowngradeVersion;
exports.isFeatureEnabled = dist_thirdwebDevSdk.isFeatureEnabled;
exports.isIncrementalVersion = dist_thirdwebDevSdk.isIncrementalVersion;
exports.resolveContractUriFromAddress = dist_thirdwebDevSdk.resolveContractUriFromAddress;
exports.toSemver = dist_thirdwebDevSdk.toSemver;
